libtenzir/src/table_slice.cpp

```diff
@@ -93,33 +95,245 @@ create_table_slice(const std::shared_ptr<arrow::RecordBatch>& record_batch,
         }
-      },
-      [&](const arrow::ListArray& la) noexcept {
-        check_col(check_col, *la.values());
-      },
-      [&](const arrow::MapArray& ma) noexcept {
-        check_col(check_col, *ma.keys());
-        check_col(check_col, *ma.items());
-      },
-      [](const arrow::Array&) noexcept {},
-    };
-    match(column, f);
-  };
-  for (const auto& column : record_batch.columns()) {
-    check_col(check_col, *column);
+        new_arrays.push_back(std::move(new_array));
+        new_fields.push_back(std::move(new_field));
+      }
+      return make_struct_array(struct_array->length(),
+                               struct_array->null_bitmap(),
+                               std::move(new_fields), std::move(new_arrays));
+    }
+    case arrow::Type::LIST: {
+      auto list_array = std::static_pointer_cast<arrow::ListArray>(array);
+      auto values = upgrade_arrays(list_array->values());
+      return check(arrow::ListArray::FromArrays(
+        *list_array->offsets(), *values, arrow::default_memory_pool(),
+        list_array->null_bitmap(), list_array->null_count()));
+    }
+    default: {
+      if (auto target = target_type_for(*array->type())) {
+        auto result = check(arrow::compute::Cast(array, target));
+        return result.make_array();
+      }
+      return array;
+    }
+  }
+  return array;
+}
+
+/// Tries to upgrade some column types to types supported by Tenzir, by
+/// recursively checking their types and casting.
+auto upgrade_record_batch(const std::shared_ptr<arrow::RecordBatch>& batch)
+  -> std::shared_ptr<arrow::RecordBatch> {
+  auto new_columns = std::vector<std::shared_ptr<arrow::Array>>{};
+  auto new_fields = std::vector<std::shared_ptr<arrow::Field>>{};
+  new_columns.reserve(batch->num_columns());
+  new_fields.reserve(batch->num_columns());
+  for (auto i = 0; i < batch->num_columns(); ++i) {
+    auto array = batch->column(i);
+    const auto& field = batch->schema()->field(i);
+    auto new_array = upgrade_arrays(array);
+    auto new_field = field;
+    if (new_array != array) {
+      new_field
+        = arrow::field(field->name(), new_array->type(), field->nullable());
+    }
+    new_columns.push_back(std::move(new_array));
+    new_fields.push_back(std::move(new_field));
+  }
+  auto new_schema = std::make_shared<arrow::Schema>(new_fields);
+  return arrow::RecordBatch::Make(std::move(new_schema), batch->num_rows(),
+                                  std::move(new_columns));
+}
+
+/// Verifies that an arrays type is supported by tenzir
+/// @param arr the array to validate
+/// @param[out] an error message on failure. If the array is invalid, `error`
+///             will have a value.
+/// @returns true if the array is supported, false otherwise. If false, `error`
+///          will have a value.
+auto verify_column(const arrow::Array& arr,
+                   std::optional<table_slice::creation_error>& error) -> bool;
```

Comment:
Why not just return a `std::optional<table_slice::creation_error>` (that could just be `std::optional<std::string>`)?